### DISCUSSION TITLE
Type inference question

### DISCUSSION THREAD

**Discussion by alycklama** [zio/zio#6207] [Q&A] [answered]
I'm experimenting with ZIO and noticed the following:

```
  def a: Either[String, Int] = Right(1)
  def b: ZIO[Any, String, Int] = ZIO.fromEither(a)
  def c: ZIO[Any, Object, Int] = // I would expect ZIO[Any, String, Int] here, but that will not compile.
    b.collect(identity _) {
      case i if i % 2 == 0 => i
    }
```

Any idea why `def c: ZIO[Any, String, Int] = ...` does not compile?

**Comment by adamgfraser** [comment, marked as answer]
If you look at the signature of `collect` it is:

```scala
trait ZIO[-R, +E, +A] {
  def collect[E1 >: E, B](e: => E1)(pf: PartialFunction[A, B]): ZIO[R, E1, B] =
    ???
}
```

You can see that the first parameter is expecting an error that is a super type of `E` that we will fail with if the `A` value does not satisfy the partial function. Normally this would be the same type as the original error type, for example a `String` here like `the number wasn't even`.

However, you have supplied the identity function as the first argument. So what type does the Scala compiler unify `String` and `Function` to? `Object` is the most specific super type.

So type inference is working correctly here you just want to supply an actual value as the first parameter instead of a function.

**Comment by alycklama** [comment]
Thank you for your reply! I totally misunderstood the API here, sorry for that. I was under the impression that the error was also to be collected (hence the identity function). However it's clear to me know that the `e: => E1` represents the fallback case, when nothing can be matched.
